pytorch_lightning/core/lightning.py

```diff
@@ -1514,32 +1517,33 @@ def to_onnx(self, file_path: str, input_sample: Optional[Tensor] = None, **kwarg
             ...     os.path.isfile(tmpfile.name)
             True
         """
-
-        if isinstance(input_sample, Tensor):
-            input_data = input_sample
-        elif self.example_input_array is not None:
-            input_data = self.example_input_array
-        else:
-            if input_sample is not None:
-                raise ValueError(
-                    f"Received `input_sample` of type {type(input_sample)}. Expected type is `Tensor`"
-                )
-            else:
+        if input_sample is None:
+            if self.example_input_array is None:
                 raise ValueError(
                     "Could not export to ONNX since neither `input_sample` nor"
                     " `model.example_input_array` attribute is set."
                 )
-        input_data = input_data.to(self.device)
+            input_sample = self.example_input_array
+
+        if isinstance(input_sample, Tensor):
+            input_sample = (input_sample,)
+        elif not (isinstance(input_sample, tuple) and all(isinstance(inp, Tensor) for inp in input_sample)):
+            raise ValueError(
+                "Could not export to ONNX since input_sample is neither a Tensor nor tuple of Tensors"
+            )
+
+        input_sample = self.transfer_batch_to_device(input_sample, self.device)
+
         if "example_outputs" not in kwargs:
             self.eval()
-            with torch.no_grad():
-                kwargs["example_outputs"] = self(input_data)
+            kwargs["example_outputs"] = self(*input_sample)
 
-        torch.onnx.export(self, input_data, file_path, **kwargs)
+        torch.onnx.export(self, input_sample, file_path, **kwargs)
 
+    @torch.no_grad()
     def to_torchscript(
         self, file_path: Optional[str] = None, method: Optional[str] = 'script',
-            example_inputs: Optional[Union[torch.Tensor, Tuple[torch.Tensor]]] = None, **kwargs
+        example_inputs: Optional[Union[Tensor, Tuple[Tensor]]] = None, **kwargs
```

Comment:
https://pytorch.org/docs/stable/jit_language_reference.html

It would be worth to extend our API to support all TorchScript supported types such as NamedTuple, etc..

Reply:
added for list and NamedTuple. Others are not supported